File: solutions/cpp/robot-simulator/1/robot_simulator.cpp
#include "robot_simulator.h"
#include <utility>

namespace robot_simulator {
std::pair<int, int> Robot::get_position() const { return {x, y}; }
Bearing Robot::get_bearing() const { return dir; }
Robot::Robot(std::pair<int, int> i_pos, Bearing i_dir) {
  x = i_pos.first;
  y = i_pos.second;
  dir = i_dir;
}
void Robot::turn_left() {
  int i = static_cast<int>(dir);
  i = i <= 0 ? 3 : i - 1;
  dir = static_cast<Bearing>(i);
}
void Robot::turn_right() {
  int i = static_cast<int>(dir);
  i = i >= 3 ? 0 : i + 1;
  dir = static_cast<Bearing>(i);
}
void Robot::advance() {
  switch (dir) {
  case Bearing::NORTH:
    y++;
    break;
  case Bearing::EAST:
    x++;
    break;
  case Bearing::SOUTH:
    y--;
    break;
  case Bearing::WEST:;
    x--;
    break;
  }
}

void Robot::execute_sequence(const std::string &str) {
  for (const auto &c : str) {
    switch (c) {
    case 'R':
      turn_right();
      break;
    case 'L':
      turn_left();
      break;
    case 'A':
      advance();
      break;
    }
  }
}

} // namespace robot_simulator

File: solutions/lua/robot-simulator/1/robot-simulator.lua
local cmds = {
  A = {
    north = function(self)
      self.y = self.y + 1
    end,
    west = function(self)
      self.x = self.x - 1
    end,
    east = function(self)
      self.x = self.x + 1
    end,
    south = function(self)
      self.y = self.y - 1
    end
  },
  R = {
    north = function(self)
      self.heading = "east"
    end,
    west = function(self)
      self.heading = "north"
    end,
    south = function(self)
      self.heading = "west"
    end,
    east = function(self)
      self.heading = "south"
    end
  },
  L = {
    north = function(self)
      self.heading = "west"
    end,
    west = function(self)
      self.heading = "south"
    end,
    south = function(self)
      self.heading = "east"
    end,
    east = function(self)
      self.heading = "north"
    end
  }
}

local function move(self, cmd)
  for i in cmd:gmatch(".") do
    cmds[i][self.heading](self)
  end
end

return function(config)
  return {
    x = config.x,
    y = config.y,
    heading = config.heading,
    move = move
  }
end
